parser/src/parser/grammar.lalrpop

```diff
@@ -119,7 +137,10 @@ BoundaryAtom: BoundaryExpr = {
     <lexpr: BoundaryAtom> "^" <num: Num_u64> => BoundaryExpr::Exp(Box::new(lexpr), num),
     "$rand" "[" <n: Num_u64> "]" => BoundaryExpr::Rand(n as usize),
     <n: Num_u64> => BoundaryExpr::Const(n),
-    <ident: Identifier> "[" <index: Num_u64> "]" => BoundaryExpr::PubInput(ident, index as usize)
+    <ident: Identifier> => BoundaryExpr::Var(ident),
+    <ident: Identifier> "[" <idx: Num_u64> "]" => BoundaryExpr::VecElem(ident, idx as usize),
+    <ident: Identifier> "[" <idx_row: Num_u64> "]" "[" <idx_col: Num_u64> "]" =>
+        BoundaryExpr::MatrixElem(ident, idx_row as usize, idx_col as usize)
```

Comment:
I think we should break this out into something reusable that we can also use for transition constraints and in the future for column groupings.

I'd prefer something like a `VectorAccess` whose index is a number or another vector access. (I haven't thought about this a ton yet though so lmk if you want to discuss)

Reply:
I've left a code sample in the latest review